onert-micro/eval-driver/TrainingDriver.cpp

```diff
@@ -100,117 +92,241 @@ int entry(int argc, char **argv)
+  }
+  else
+  {
+    throw std::runtime_error("Unknown commands number\n");
   }
 
-  auto test_data_u8 = reinterpret_cast<char *>(test_data);
+  DataBuffer circle_model = readFile(circle_model_path);
+  DataBuffer wof_data;
+  // If defined wof file
+  if (wof_file_path != nullptr)
+    wof_data = readFile(wof_file_path);
+
+  // Configure training mode
+  onert_micro::OMConfig config;
+
+  // Save model size and model ptr in config
+  config.model_size = circle_model.size();
+  config.model_ptr = circle_model.data();
+
+  // If defined wof file
+  if (wof_file_path != nullptr)
+    config.wof_ptr = nullptr;
+
+  // Set user defined training settings
+  const uint32_t training_epochs = 50;
+  const float lambda = 0.001f;
+  const uint32_t BATCH_SIZE = 32;
+  const uint32_t INPUT_SIZE = 180;
+  const uint32_t OUTPUT_SIZE = 4;
+  const uint32_t num_train_layers = 0;
+  const onert_micro::OMLoss loss = onert_micro::CROSS_ENTROPY;
+  const onert_micro::OMTrainOptimizer train_optim = onert_micro::ADAM;
+  const float beta = 0.9;
+  const float beta_squares = 0.999;
+  const float epsilon = 1e-07;
+
+  config.train_mode = true;
+  {
+    onert_micro::OMTrainingContext train_context;
+    train_context.batch_size = BATCH_SIZE;
+    train_context.num_of_train_layers = num_train_layers;
+    train_context.lambda = lambda;
+    train_context.loss = loss;
+    train_context.optimizer = train_optim;
+    train_context.beta = beta;
+    train_context.beta_squares = beta_squares;
+    train_context.epsilon = epsilon;
+
+    config.training_context = train_context;
+  }
 
-  printf("Run test dataset:\n");
-  for (int i = 0; i < num_test_data_samples; ++i)
+  // Create training interpreter and import models
+  onert_micro::OMTrainingInterpreter train_interpreter;
+  train_interpreter.importTrainModel(circle_model.data(), config);
+
+  // Temporary buffer to read input data from file using BATCH_SIZE
+  float training_input[BATCH_SIZE * INPUT_SIZE];
+  float training_target[BATCH_SIZE * OUTPUT_SIZE];
+  // Note: here test size used with BATCH_SIZE = 1
+  float test_input[INPUT_SIZE];
+  float test_target[OUTPUT_SIZE];
+  std::vector<float> accuracy_v;
+  std::vector<float> cross_entropy_v;
+  float max_accuracy = std::numeric_limits<float>::min();
+
+  for (uint32_t e = 0; e < training_epochs; ++e)
   {
-    auto input_data = reinterpret_cast<char *>(interpreter.allocateInputTensor(0));
+    // Run train for current epoch
+    std::cout << "Run training for epoch: " << e + 1 << "/" << training_epochs << "\n";
+    config.training_context.num_epoch = e + 1;
+    uint32_t num_steps = num_train_data_samples / BATCH_SIZE;
+    for (int i = 0; i < num_steps; ++i)
+    {
+      uint32_t cur_batch_size = std::min(BATCH_SIZE, num_train_data_samples - BATCH_SIZE * i - 1);
+      cur_batch_size = std::max(1u, cur_batch_size);
+
+      config.training_context.batch_size = cur_batch_size;
+      config.training_context.num_step = i + 1;
```

Comment:
AFAIU, num_step for ADAM optimizer should not be reset by each epoch. 
```suggestion
      config.training_context.num_step++;
```

Reply:
or 
```suggestion
      config.training_context.adam_step++;
```

Reply:
I see, I will change it